keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,642 @@
+```
+
+### Existing Struct
+https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go
+
+https://github.com/kubernetes/kubernetes/blob/7e7bc6d53b021be6fe3d5a1125a990913b7a9028/pkg/apis/core/types.go#L2062-L2094
+
+```go
+// Probe describes a health check to be performed against a container to determine whether it is
+// alive or ready to receive traffic.
+type Probe struct {
+	// The action taken to determine the health of a container
+	ProbeHandler
+	// Length of time before health checking is activated.  In seconds.
+	// +optional
+	InitialDelaySeconds int32
+	// Length of time before health checking times out.  In seconds.
+	// +optional
+	TimeoutSeconds int32
+	// How often (in seconds) to perform the probe.
+	// +optional
+	PeriodSeconds int32
+	// Minimum consecutive successes for the probe to be considered successful after having failed.
+	// Must be 1 for liveness and startup.
+	// +optional
+	SuccessThreshold int32
+	// Minimum consecutive failures for the probe to be considered failed after having succeeded.
+	// +optional
+	FailureThreshold int32
+	// Optional duration in seconds the pod needs to terminate gracefully upon probe failure.
+	// The grace period is the duration in seconds after the processes running in the pod are sent
+	// a termination signal and the time when the processes are forcibly halted with a kill signal.
+	// Set this value longer than the expected cleanup time for your process.
+	// If this value is nil, the pod's terminationGracePeriodSeconds will be used. Otherwise, this
+	// value overrides the value provided by the pod spec.
+	// Value must be non-negative integer. The value zero indicates stop immediately via
+	// the kill signal (no opportunity to shut down).
+	// This is a beta field and requires enabling ProbeTerminationGracePeriod feature gate.
+	// +optional
+	TerminationGracePeriodSeconds *int64
+}
+```
+
+### Existing Behavior
+
+Taking the Seconds fields, in order of the struct.
+InitialDelaySeconds
+TimeoutSeconds
+PeriodSeconds
+
+#### Defaulting logic
+ - InitialDelaySeconds, has no defaulting logic.
+   Therefore it would get the golang defaulting logic, and default to 0.
+ - TimeoutSeconds, defaults to 1 second if unset (which includes the explicitly set to 0 state). This is the last ending time before the timeout fails. Defaulting, thus, results in slow failures (when one considers a full second of time in cpu time).
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L224-L226
+
+```go
+ 	if obj.TimeoutSeconds == 0 {
+		obj.TimeoutSeconds = 1
+	}
+```
+
+ - Period Seconds is the biggest hurdle in charging and overall cost to the cluster, but also the most useful if set correctly.
+PeriodSeconds defaults to 10 seconds if unset (which includes the explicitly set to 0 state). Fast failure detections are thus blocked with existing defaults.
```

Comment:
I would expect the new fields’ presence to block the defaulting on the old field.  That preserves user intent and is forward compatible for older objects.

I was surprised to see this approach taken (assuming we can’t skip defaulting when a second field is present), is there a discussion with a reviewer (i didn’t see this discussed in the doc text) pointing in that direction?

If we preserve the base duration defaulting we can’t change that in the future without a third field, barring a really good argument I would argue preserving the old defaulting is user hostile and confusing (and once done, is even worse to undo later).

Reply:
At minimum I need to see this question resolved prior to the api change merging, but better to do it here.

Reply:
note: this is in the existing behavior section 
See above for proposed change and examples copied here:
```
`periodSeconds: 1` and `periodMilliseconds: -500` would be 0.5s / 500ms. (*** To reduce un-necessary resource usage, because periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer becoming 1second after success.)

 More generally, the effective period value would be = `periodSeconds` + `periodMilliseconds`.

 There are a few corner cases around default (effective period) values:

 `periodSeconds: 0` and `periodMilliseconds: 500` would be 10.5s / 10500ms (as 0 => 10 for `periodSeconds` via [defaults.go](https://github.com/kubernetes/kubernetes/blob/3b13e9445a3bf86c94781c898f224e6690399178/pkg/apis/core/v1/defaults.go#L213-L215))
```
Is this acceptable? .. cheers..

Reply:
Changing periodSeconds defaulting behavior when periodMilliseconds is set, and the gate is on/enabled in a release of kubelet that supports it, and periodSeconds is set to zero based on a user "intent" understanding to override the existing defaulting behavior if the gate is on in a newer release and back to 10 seconds if run with gate off or deployed to an older release.. I believe that would be a new proposal.

Reply:
if the reviewers feel overriding of default behavior with the gate on and when mills are present.. Works for me :-)

Reply:
I think we landed on not changing the default behavior because we assumed that would be an [incompatible change](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility) and thus a no-go. But if that's not the case, then agree that overriding the 0->10 default would be ideal.

Reply:
I do not believe that changing defaulting behavior in the presence of a new optional field is incompatible according to our definition.  The reason being is that we change behavior **when we add new optional fields**.  For instance, pod readiness gates delay readiness indefinitely.  The "default" behavior (scare quotes deliberate) is that pods go ready as soon as their readiness test passes, but adding readiness gates (which old clients aren't aware of) will change that behavior in a way that an old client can't observe but is nevertheless different.

The presence of an additional field is sufficient justification to change the observed behavior of an existing field across API versions, even if an older client can't observe *why* the behavior changed.

Server-defaulting changes would be a subset of general API behavior changes.  A client that doesn't provide the readiness probe duration is accepting the server default, and they generally expect the observed behavior of that default to not change across new versions **absent another input**.  That additional input is generally new API fields, but occasionally administrator configuration or a security fix will also alter the behavior.  Sometimes the observed behavior is incorrect - we might alter the server defaulted value if a serious bug is the consequence of the current defaulted value.

So, unless there is a very strong argument (will check with api-reviewers), **the presence of a new field may alter the defaulting of an existing field as long as the user can reason about the interaction of those fields and validation is correctly handled**.

Reply:
nod.. 

on the confusing users issue I can be swayed both ways .. to use the field as an offset and boolean override of the 10second default.. or leave it. Key will, IMO, be to provide explicit documentation examples to eliminate all confusion and proper tests to ensure it works as advertised.